src/principles.rs

```diff
@@ -0,0 +1,117 @@
+use codec::{Decode, Encode};
+use rstd::collections::btree_set::BTreeSet;
+
+/// The principal type to which a permission can be assigned.
+#[derive(Encode, Decode, Eq, PartialEq, Ord, PartialOrd, Clone, Debug)]
+pub enum BasePrincipal<AccountId, GroupId> {
+    Account(AccountId),
+    GroupMember(GroupId),
+}
+
+/// The principal type to which entity permissions can be assigned.
+#[derive(Encode, Decode, Eq, PartialEq, Ord, PartialOrd, Clone, Debug)]
+pub enum EntityPrincipal<AccountId, GroupId> {
+    Base(BasePrincipal<AccountId, GroupId>),
+    Owner,
```

Comment:
Should be called `EntityOwner` as in `DerivedPrincipal`?
```suggestion
    EntityOwner,
```

Reply:
outdated